# Integrantes



*   Luis Alejandro Montoya
*   Juan Sebastian Villada
*   Christian Daniel España



In [1]:
import numpy as np

In [7]:
# Base de datos de productos y precios
base_de_datos = {
            "Pantalon": 100000,
            "Camiseta": 50000,
            "Gafas": 25000,
            "Gorra": 20000,
            "Zapatos": 250000
            # Agrega más productos y precios según sea necesario
        }


class Pedido:

    """
    Clase Pedido que ayuda a tomar los pedidos enviados por el usuario y calcula el total de la compra

    -----------

    Atributos
    ----------
    numero_pedido : int
        Identificador del pedido
    cliente : str
        Nombre del cliente quien realiza el pedido
    fecha : str
        Fecha de creación del pedido
    productos : dict
        Diccionario de productos seleccionados por el usuario


    Metodos
    -------
    mostrar_detalle()
        Muestra los productos y la cantidad de ellos

    agregar_producto(producto, cantidad)
        Agrega nuevos productos o actualiza los exitentes

    eliminar_producto(producto)
        Elimina el producto seleccionado

    calcular_total()
        Calcula el total de los productos realizando una busqueda de los precios en la base de datos
    """
    def __init__(self, numero_pedido: int, cliente: str, fecha: str, productos: dict):
      self.informacion_pedido = {
          'numero_pedido': numero_pedido,
          'cliente': cliente,
          'fecha': fecha,
          'productos': dict(productos),
          'total' : 0,
      }

      """
      Parametros
      ----------
      numero_pedido : int
          Identificador del pedido
      cliente : str
          Nombre del cliente quien realiza el pedido
      fecha : str
          Fecha de creación del pedido
      productos : dict
          Diccionario de productos seleccionados por el usuario
      """



    def mostrar_detalle(self):

      """
      Esta función muestra el detalle de la compra, los productos seleccionados por el usuario y el total
      """

      print(f"Número de Pedido: {self.informacion_pedido['numero_pedido']}")
      print(f"Cliente: {self.informacion_pedido['cliente']}")
      print(f"Fecha: {self.informacion_pedido['fecha']}")
      print("Productos:")
      for producto, cantidad in self.informacion_pedido['productos'].items():
          print(f" - {producto}: {cantidad}")
      self.calcular_total()
      print()
      print(f"Total: COP {self.informacion_pedido['total']}")

    def agregar_producto(self, producto, cantidad):

      """
      Esta función se encarga de agregar nuevos productos a la lista de compra
      """

      #Se realiza la busqueda del producto en la base inicial de productos, si lo encuentra suma la cantidad sino lo agrega
      if producto in self.informacion_pedido['productos']:
          self.informacion_pedido['productos'][producto] += cantidad
      else:
          self.informacion_pedido['productos'][producto] = cantidad

    def eliminar_producto(self, producto):

      """
      Esta función elimina el producto de la lista de compra
      """

      #Valida si el producto se encuentra en la base inicial de productos de ser asi lo elimina
      if producto in self.informacion_pedido['productos']:
          del self.informacion_pedido['productos'][producto]
      else:
          print(f"El producto '{producto}' no se encuentra en el pedido.")

    def calcular_total(self):

      """
      Es la función que realiza el calculo del total a pagar tomando el valor de cada item y las cantidades
      """

      total = 0
      # Se recorre el diccionario que contiene los productos seleccionados por el usuario
      for producto, cantidad in self.informacion_pedido['productos'].items():
          # Si el producto se encuentra en la base de datos, se actualiza el total a pagar multiplicando el precio alojado en la base de datos con la cantidad que se le pasa por parametro
          if producto in base_de_datos:
              precio = base_de_datos[producto]
              total += precio * cantidad
          else:
            # En caso de que el producto no se encuentre en la base de datos, el programa permite incluir el nuevo item con su valor respectivo.
            print(f"*** El producto {producto} seleccionado por el usario no existe en la base de datos ***")
            precio = int(input(f"Ingrese el precio del nuevo producto: {producto} \t"))
            base_de_datos[producto] = precio
            total += precio * cantidad

      self.informacion_pedido['total'] = total
      return total


In [8]:
# Crear un pedido
productos_iniciales = {"Pantalon": 1, "Camiseta": 1}
# Implementación de la clase Pedido
pedido1 = Pedido(1, "Laura", "2023-08-05", productos_iniciales)

In [9]:
# Mostrar detalles del pedido
pedido1.mostrar_detalle()

Número de Pedido: 1
Cliente: Laura
Fecha: 2023-08-05
Productos:
 - Pantalon: 1
 - Camiseta: 1

Total: COP 150000


In [10]:
# Eliminar el item "Pantalon"
pedido1.eliminar_producto("Pantalon")
# Mostrar nuevamente los detalles de la compra
pedido1.mostrar_detalle()


Número de Pedido: 1
Cliente: Laura
Fecha: 2023-08-05
Productos:
 - Camiseta: 1

Total: COP 50000


In [11]:
# Agregar productos
pedido1.agregar_producto("aretes", 1)
# Mostrar productos nuevamente
pedido1.mostrar_detalle()

Número de Pedido: 1
Cliente: Laura
Fecha: 2023-08-05
Productos:
 - Camiseta: 1
 - aretes: 1
*** El producto aretes seleccionado por el usario no existe en la base de datos ***
Ingrese el precio del nuevo producto: aretes 	19999

Total: COP 69999


In [13]:
class PedidoDescuento(Pedido):

  """
    Clase PedidoDescuento que ayuda a realizar un descuento de los productos seleccionados la cual hereda de la clase Pedido

    -----------

    Atributos
    ----------
    numero_pedido : int
        Identificador del pedido
    cliente : str
        Nombre del cliente quien realiza el pedido
    fecha : str
        Fecha de creación del pedido
    productos : dict
        Diccionario de productos seleccionados por el usuario
    descuento : int
        valor con el que se aplicará el descuento


    Metodos
    -------
    mostrar_detalle()
        Muestra los productos y la cantidad de ellos el total con y sin descuento
  """


  def __init__(self, numero_pedido: int, cliente: str, fecha: str, productos: dict, descuento: int):


    super().__init__(numero_pedido, cliente, fecha, productos)
    self.descuento = descuento

    """
    Atributos
    ----------
    numero_pedido : int
        Identificador del pedido
    cliente : str
        Nombre del cliente quien realiza el pedido
    fecha : str
        Fecha de creación del pedido
    productos : dict
        Diccionario de productos seleccionados por el usuario
    descuento : int
        valor con el que se aplicará el descuento
    """

  # Con la creación de esta funcion estamos cumpliendo unos de los pilares de la POO el cual es el polimorfismo
  def mostrar_detalle(self):

    # Se invoca el metodo mostrar_detalle de la clase madre Pedido
    super().mostrar_detalle()
    # Se invoca el metodo calcular_total de la clase madre Pedido
    super().calcular_total()

    # Se calcula el descuento para los productos seleccionados
    descuento_total = self.informacion_pedido['total'] * (self.descuento / 100)
    # Se resta el descuento a la variable total
    self.informacion_pedido['total'] -= descuento_total

    print(f"Descuento aplicado: {self.descuento}%")
    print(f"Total con descuento: COP {self.informacion_pedido['total']}")


In [14]:
# Se instancia la clase hija
pedido_descuento = PedidoDescuento(numero_pedido=1, cliente="Laura", fecha="2023-08-07", productos=productos_iniciales, descuento=50)

# Mostrar los detalles del pedido con descuento
pedido_descuento.mostrar_detalle()

Número de Pedido: 1
Cliente: Laura
Fecha: 2023-08-07
Productos:
 - Pantalon: 1
 - Camiseta: 1

Total: COP 150000
Descuento aplicado: 50%
Total con descuento: COP 75000.0


Pertinencia de la clase en el ámbito, ¿Por qué será útil esta herramienta y cómo ayudará a mejorar el negocio?

Respuesta:

El programa es util por que es una herramienta que ayuda al manejo del inventario por medio de la gestión de la base de datos, ademas permite realizar de manera eficiente el calculo de la venta asociada a los pedidos que son despachados desde la tienda, toda vez que permite automatizar y estandarizar tareas como añadir nuevos items al pedido, eliminar items, listar los productos de la compra, realizar descuentos  y calcular el total a pagar.



#Ejercicio 2

#a.)

In [15]:
# Coeficientes de las ecuaciones
coefficients = np.array([[2, 3, -1, 4, 5],
                         [1, -2, 4, -3, 1],
                         [3, 2, -3, 5, -2],
                         [4, 1, -2, 3, 2],
                         [1, 1, 1, 1, 1]])

# Términos independientes
constants = np.array([15, 6, 11, 8, 7])

In [16]:
# Resuelve el sistema de ecuaciones utilizando la función np.linalg.solve():
solution = np.linalg.solve(coefficients, constants)

print("Solución:\n")
print("x =", round(solution[0], 2))
print("y =", round(solution[1], 2))
print("z =", round(solution[2], 2))
print("w =", round(solution[3], 2))
print("v =", round(solution[4], 2))

Solución:

x = -1.68
y = -7.87
z = 5.13
w = 10.0
v = 1.42


#b.)

In [17]:
from scipy.special import factorial

def euler(n):
  '''
  Esta funcion realiza el calculo del valor del número de Euler

  Parametros de entrada:
    n (int): Numero entero

  Retorna:
  int: Valor real que corresponde la sumatoria de la funcion de Euler hasta 1 hasta n

  '''
  # Se crea un array de longitud n
  l = np.arange(n)
  # Se calcula el factorial para cada elemento del array
  factoriales = factorial(l)
  # Se realiza la sumatoria sum((1) / (factoriales)) y se retarna el valor calculado
  return np.sum(((1) / (factoriales) ) )

In [18]:
# Se evalua la función construida con el valor de 4 y no se aproxima al valor deseado
euler(4)

2.6666666666666665

In [19]:
# Con el valor de 10 en adelante se logra exitosamente el valor deseado
euler(10)

2.718281525573192

In [20]:
euler(10000)

2.718281828459045


**Solución:**


El valor deseado se alcanza con la primera potencia de 10, es decir 10.

#c.)

In [21]:
def calculate_polynomial(matrix):
  '''
  Esta funcion realiza el calculo del polinomio caracteristico para la matriz cuadrada

  Parametros de entrada:
    matrix (numpy.ndarray): Matriz cuadrada

  Retorna:
  numpy.ndarray: Descripción del valor retornado

  '''
  # Se calcula los eigenvalores de la matriz de entrada
  eigenvalores = np.linalg.eigvals(matrix)

  # Se retorna los eigenvalores de la matriz
  return eigenvalores

In [22]:
# Implementación de la matriz cuadrada
matriz = np.array([[2, 1, 1],
                    [4, 3, 2],
                    [1, 1, 2]])

vals = calculate_polynomial(matriz)
print(np.round(vals,2))

[5.45 0.55 1.  ]


Luego de obtener los eigenvalores para lambda λ, se realiza el calculo:

(λ - 5.45)(λ - 0.550)(λ - 1) = 0


**Solución:**

λ³ - 7.000 λ² + 8.977 λ - 2.997 = 0